plugins/org.eclipse.sirius.diagram.ui/src-diag/org/eclipse/sirius/diagram/ui/tools/internal/actions/pinning/PinElementsAction.java

```diff
@@ -187,48 +170,57 @@ private void updateActionState(ISelection selection) {
      * Add listener of the pin/unpin feature of the last selected element selection (selectionLastElement) if present.
      */
     private void addPinListener() {
-        selectionLastElement.ifPresent(elem -> {
-            TransactionalEditingDomain domain = TransactionUtil.getEditingDomain(elem);
-            if (domain != null) {
-                pinListener = new PinHelper().createPinListener(elem, () -> {
-                    EclipseUIUtil.displayAsyncExec(this::updatePinState);
-                });
-                domain.addResourceSetListener(pinListener);
-
-                Resource resource = elem.eResource();
-                if (resource != null) {
-                    var permissionAuthoriry = PermissionAuthorityRegistry.getDefault().getPermissionAuthority(resource.getResourceSet());
-                    permissionAuthoriry.addAuthorityListener(authorityListener);
+        if (selectionElements != null) {
+            selectionElements = selectionElements.keySet().stream().collect(Collectors.toMap(Function.identity(), (elem) -> {
+                TransactionalEditingDomain domain = TransactionUtil.getEditingDomain(elem);
+                if (domain != null) {
+                    ResourceSetListener pinListener = new PinHelper().createPinListener(elem, () -> {
+                        EclipseUIUtil.displayAsyncExec(this::updatePinState);
+                    });
+                    domain.addResourceSetListener(pinListener);
+
+                    Resource resource = elem.eResource();
+                    if (resource != null) {
+                        var permissionAuthoriry = PermissionAuthorityRegistry.getDefault().getPermissionAuthority(resource.getResourceSet());
+                        permissionAuthoriry.addAuthorityListener(authorityListener);
+                    }
```

Comment:
it seems that you do not need to add and remove as many authority listener as selected elements

Reply:
Ok, I now have one listener per ResourceSet (only one in most cases)